src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -67,13 +67,35 @@ public GeneratorDriver RunGeneratorsAndUpdateCompilation(Compilation compilation
 
         public GeneratorDriver AddGenerators(ImmutableArray<ISourceGenerator> generators)
         {
-            (var filteredGenerators, var incrementalGenerators) = GetIncrementalGenerators(generators, SourceExtension);
-            var newState = _state.With(sourceGenerators: _state.Generators.AddRange(filteredGenerators),
+            var incrementalGenerators = GetIncrementalGenerators(generators, SourceExtension);
+            var newState = _state.With(sourceGenerators: _state.Generators.AddRange(generators),
                                        incrementalGenerators: _state.IncrementalGenerators.AddRange(incrementalGenerators),
-                                       generatorStates: _state.GeneratorStates.AddRange(new GeneratorState[filteredGenerators.Length]));
+                                       generatorStates: _state.GeneratorStates.AddRange(new GeneratorState[generators.Length]));
             return FromState(newState);
         }
 
+        public GeneratorDriver ReplaceGenerators(ImmutableArray<ISourceGenerator> generators)
+        {
+            var incrementalGenerators = GetIncrementalGenerators(generators, SourceExtension);
+            var states = ArrayBuilder<GeneratorState>.GetInstance(generators.Length);
+
+            foreach (var generator in generators)
+            {
+                var existingIndex = _state.Generators.IndexOf(generator);
```

Comment:
Is there something we should document for this call publicly? As in, recommendations for generator authors about equality of their types? Always advise them to use reference equality?

Reply:
@chsienki Did we make any strong documentation anywhere here? Note that RemoveGenerators also makes similar assumptions to here, so any assumption isn't new.

Reply:
I don't think we ever doc'd it explicitly. But yeah, we should probably advise them to always use ref equality